File: solutions/code-carl/ch08_greedy/t7_lc1005_largest_sum_after_k_negations.py
"""LeetCode 1005. Maximize Sum Of Array After K Negations
https://leetcode.cn/problems/maximize-sum-of-array-after-k-negations/
"""
import sys
from typing import List


class Solution:
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        nums.sort()
        ans: int = 0
        min_abs: int = sys.maxsize

        for i, num in enumerate(nums):
            min_abs = min(min_abs, abs(num))
            if num < 0 < k:
                k -= 1
                nums[i] = -num
            ans += nums[i]

        return ans if k % 2 == 0 else ans - 2 * min_abs


if __name__ == "__main__":
    solution = Solution()
    assert solution.largestSumAfterKNegations([4, 2, 3], 1) == 5
    assert solution.largestSumAfterKNegations([3, -1, 0, 2], 3) == 6
    assert solution.largestSumAfterKNegations([2, -3, -1, 5, -4], 2) == 13
    print("All tests passed.")

File: solutions/java/code-carl/ch08_greedy/T7_LC1005_LargestSumAfterKNegations.java
/**
 * Practice template for:
 * - LeetCode 1005. Maximize Sum Of Array After K Negations
 *   https://leetcode.cn/problems/maximize-sum-of-array-after-k-negations/
 */
import java.util.Arrays;

public class T7_LC1005_LargestSumAfterKNegations {
    public int largestSumAfterKNegations(int[] nums, int k) {
        int minAbs = Integer.MAX_VALUE;
        int sum = 0;
        Arrays.sort(nums);

        for (int i = 0; i < nums.length; i++) {
            minAbs = Math.min(minAbs, Math.abs(nums[i]));
            if (k > 0 && nums[i] < 0) {
                k--;
                nums[i] = -nums[i];
            }
            sum += nums[i];
        }

        if (k % 2 == 1) {
            sum -= 2 * minAbs;
        }
        return sum;
    }
}
